### PR TITLE
[spirv]  Migration to v2 infrastructure

### PR DESCRIPTION
**Pull request by ehsannas**
This change is to migrate SPIR-V code generation to a new infrastructure that include an in-memory representation of SPIR-V and allows for easier traversals on the in-memory SPIR-V module. Also a large number of cleanups and refactoring of the existing SPIR-V backend code.

Does not touch non-SPIRV code.

### CODE REVIEW

File: lib/Support/YAMLTraits.cpp [hunk 1]
```diff
@@ -10,7 +10,6 @@
 #include "llvm/Support/YAMLTraits.h"
 #include "llvm/ADT/SmallString.h"
 #include "llvm/ADT/Twine.h"
-#include "llvm/Support/Casting.h"
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply lib/Support/YAMLTraits.cpp]: oops. Fixed.

**Comment by alelenv** [comment]
Lei/Ehsan, any idea when this will be merged?

**Comment by ehsannas** [comment]
@alelenv We are still running tests. Due to the size of this change and that rolling it back would not be easy, we want to make sure there are no regressions.

**Comment by alelenv** [comment]
@ehsannas : Ok Thanks!

**Pull request closed without merging**